core/src/main/java/org/elasticsearch/common/inject/internal/MoreTypes.java

```diff
@@ -176,84 +176,89 @@ public static Type canonicalize(Type type) {
      * Returns true if {@code a} and {@code b} are equal.
      */
     public static boolean equals(Type a, Type b) {
-        if (a == b) {
-            // also handles (a == null && b == null)
-            return true;
+        while (true) {
+            if (a == b) {
+                // also handles (a == null && b == null)
+                return true;
 
-        } else if (a instanceof Class) {
-            // Class already specifies equals().
-            return a.equals(b);
+            } else if (a instanceof Class) {
+                // Class already specifies equals().
+                return a.equals(b);
 
-        } else if (a instanceof ParameterizedType) {
-            if (!(b instanceof ParameterizedType)) {
-                return false;
-            }
+            } else if (a instanceof ParameterizedType) {
+                if (!(b instanceof ParameterizedType)) {
+                    return false;
+                }
 
-            // TODO: save a .clone() call
-            ParameterizedType pa = (ParameterizedType) a;
-            ParameterizedType pb = (ParameterizedType) b;
-            return Objects.equals(pa.getOwnerType(), pb.getOwnerType())
+                // TODO: save a .clone() call
+                ParameterizedType pa = (ParameterizedType) a;
+                ParameterizedType pb = (ParameterizedType) b;
+                return Objects.equals(pa.getOwnerType(), pb.getOwnerType())
                     && pa.getRawType().equals(pb.getRawType())
                     && Arrays.equals(pa.getActualTypeArguments(), pb.getActualTypeArguments());
 
-        } else if (a instanceof GenericArrayType) {
-            if (!(b instanceof GenericArrayType)) {
-                return false;
-            }
+            } else if (a instanceof GenericArrayType) {
+                if (!(b instanceof GenericArrayType)) {
+                    return false;
+                }
 
-            GenericArrayType ga = (GenericArrayType) a;
-            GenericArrayType gb = (GenericArrayType) b;
-            return equals(ga.getGenericComponentType(), gb.getGenericComponentType());
+                GenericArrayType ga = (GenericArrayType) a;
+                GenericArrayType gb = (GenericArrayType) b;
+                b = gb.getGenericComponentType();
+                a = ga.getGenericComponentType();
```

Comment:
The actual update. If the if/else branching is "perfect", meaning no `return null` after the last one, then there's no reason to do a `continue`